feat: Add first files

Salv-23 opened the pull request:
This pull request will add the first files for the musicbrainz client.

Salv-23 committed:
feat: Add first files

hector-sab commented:
**suggestion(project-structure):** I would put everything related to the REST API call to musicbrainz in its own class. So you just need one single client to query whatever you need.

hector-sab commented:
This is the structure I would recommend we followed for this project right now.
```
 metamusic.git
├── .gitignore
├── metamusic
│   ├── __init__.py
│   ├── __main__.py
│   ├── cli.py
│   └── clients
│       ├── __init__.py
│       └── musicbrainz.py
├── pyproject.toml
└── setup.py
```

Note that everything related to calling the musicbrainz api should be contained in `musicbrainz.py`.

For example:
```python
# metamusic/clients/musicbrainz.py

class MusicBrainzClient:
    base_url = ...
    def __init__(self, ...):
        ...
    def find_artist(self, name: str, minimum_acceptable_score: int = 100):
        ....
    def get_artist_discography(self, artist_id: str):
        ...
```

Let me know if you have any doubts about this or if you disagree :)

Salv-23 committed:
feat: Add artist query to MusicMetadata Class

hector-sab commented on metabrainz.py:
```diff
@@ -0,0 +1,72 @@
+# --------------------------------------------------
+if __name__ == "__main__":
+    arguments = get_args()
+    artist_argument = arguments.artist
+    main(artist_argument)
```
**suggestion:** Try not to create variables that are only used once and right after its definition. Pass `arguments.artis` directly to main.

Salv-23 replied:
Solved.

hector-sab commented on metabrainz.py:
```diff
@@ -0,0 +1,72 @@
+
+
+# --------------------------------------------------
+def main(artist: list[str]) -> dict:
```
**question:** Why a list of artists?

Btw, then it is clearer if you use `artists` to reflect that it accepts several of them.

hector-sab commented on metabrainz.py:
```diff
@@ -0,0 +1,72 @@
+    release_groups_endpoint = "http://musicbrainz.org/ws/2/release-group/"
+    releases_endpoint = "http://musicbrainz.org/ws/2/release/"
+
+    def __init__(self, artist: list[str]):
+        self.artist = artist
```
**question:** Why do you need to know the artist here?

Wondering if it would make sense a more general client, where you can simply query whatever you need for any artist at any time.

hector-sab commented on metabrainz.py:
```diff
@@ -0,0 +1,72 @@
+            if artist["score"] == 100:
+                return artist
+
+    def query_artist(self) -> dict:
```
**suggestion:** Make the client so that the class doesn't need to be initialized with an artist and you can query any artist at any given time.

Related to [this](https://github.com/Salv-23/Metamusic/pull/1/commits/cc88342dbf2ac7ea8c9b245a47bacc9289c6bc63#r985263672)

hector-sab commented on metabrainz.py:
```diff
@@ -0,0 +1,72 @@
+
+    def pick_artist(self, artist_list):
+        for artist in artist_list:
+            if artist["score"] == 100:
```
**suggestion:** Leave a `TODO` comment to remember that you need to deal with artist names that don't return a 100 score result.

Alternatively raise a not implemented error whenever this case happens. Something like

```python
def pick_artists(artists: List[str]):
    artist = None
    for possible_artist in artists:
        if not artists:
            artist = possible_artist
        elif artist["score"] < possible_artist["score"]:
            artist = possible_artist
    if artist and artist["score"] != 100:
        raise raise NotImplementedError("Selection of other artists with lower score than 100 not implemented yet")
    return artist
```

Salv-23 replied:
Changed the class method to look like the one suggested. 

**suggestion**: Only deal with artist that have a score of 100 or raise IndexError/ValueError for non existing match.

musicbrainz API will return an empty list of artists if no close match found, otherwise it will return a list of artist and automatically assign a score of 100 to the first result. Here's an example result when querying "coldplai"
```python
(Pdb) l
 47             artist = "%20".join(artist)
 48             artist_query = f"{cls.artist_endpoit}?query={artist}&fmt=json"
 49             response = requests.get(artist_query)
 50             artist_list = response.json()["artists"]
 51             breakpoint()
 52  ->         return artist_list

(Pdb) pp(response.json())
{'artists': [], 'count': 0, 'created': '2022-10-07T16:51:14.545Z', 'offset': 0}
```
And here's the response for querying "tucanes":
```python
(Pdb) pp(response.json())
{'artists': [{'id': 'f1802d84-5885-4cc0-81c6-c4823510ebf4',
              'life-span': {'ended': None},
              'name': 'Los Tucanes de Tijuana',
              'score': 100,
              'sort-name': 'Tucanes de Tijuana, Los',
              'tags': [{'count': 1, 'name': 'corrido'},
                       {'count': 1, 'name': 'norteño'},
                       {'count': 1, 'name': 'cumbias'},
                       {'count': 1, 'name': 'regional mexicano'},
                       {'count': 1, 'name': 'banda'},
                       {'count': 1, 'name': 'ballad'}],
              'type': 'Group',
              'type-id': 'e431f5f6-b5d2-343d-8b36-72607fffb74b'},
             {'area': {'id': '01448ddc-6ee3-3fa4-b136-507d984e31ee',
                       'life-span': {'ended': None},
                       'name': 'Guatemala',
                       'sort-name': 'Guatemala',
                       'type': 'Country',
                       'type-id': '06dd0ae4-8c74-30bb-b43d-95dcedf961de'},
              'begin-area': {'id': '01448ddc-6ee3-3fa4-b136-507d984e31ee',
                             'life-span': {'ended': None},
                             'name': 'Guatemala',
                             'sort-name': 'Guatemala',
                             'type': 'Country',
                             'type-id': '06dd0ae4-8c74-30bb-b43d-95dcedf961de'},
              'country': 'GT',
              'id': '15da4d9d-38fd-4845-9f69-18fca559ee3a',
              'life-span': {'begin': '1995-05-12', 'ended': None},
              'name': 'Internacionales Tucanes De Guatemala',
              'score': 90,
              'sort-name': 'Internacionales Tucanes De Guatemala',
              'type': 'Orchestra',
              'type-id': 'a0b36c92-3eb1-3839-a4f9-4799823f54a5'}],
 'count': 2,
 'created': '2022-10-07T17:00:13.710Z',
 'offset': 0}
```

hector-sab commented on metabrainz.py:
```diff
@@ -0,0 +1,72 @@
+    def query_artist(self) -> dict:
+        endpoint = MusicMetadata.artist_endpoit
+        artist_query = "%20".join(self.artist)
+        query = f"{endpoint}?query={artist_query}&limit=3&fmt=json"
```
**question:** Why limit the results to three artists?

Salv-23 replied:
So far musicbrainz API returns a list of artist and assigns a score of 100 to the first returned item in the list, after that, the scores goes down to 80-60 on average for the second result and <50 for the remaining items. If there isn't any match with a score of 100 it returns an empty list instead. That's why I figured out that it wasn't necessary to iterate over more items than 3.

hector-sab commented on metabrainz.py:
```diff
@@ -0,0 +1,72 @@
+        response = requests.get(query)
+        artist_list = response.json()["artists"]
+        my_artist = self.pick_artist(artist_list)
+        pprint(my_artist)
```
**suggestion:** Use the `logging` library instead.

Salv-23 replied:
**question**: What would be the purpose of using the `logging` library? 

I only use `pprint` for testing purposes. maybe it'll be a good idea to remove those pprint statements when comitting? The final client won't print the response.

hector-sab replied:
The idea of having logs in your application is that you can debug or get useful runtime information if needed. For example, if you know that your program is going to execute 3 different functions you could log a message whenever the function is executed. Then if something fails you can change the logger level to debug and see all the messages and double check that your program is in fact executing all three functions.

Example:
Imagine that I want main to execute function a, and function a executes function b, and b executes c. But the program is not actually executing everything. A way of debugging this is by adding print statements, other is using breakpoints, but what happens if is someone else that is executing your program? How would you be able to help them without having to jump into a call? You could ask them to set the logger level to debug, and that they send you all the logs your program is generating so that you can inspect it in hopes you find something weird.

In this example, you would note in the logs that function c is never executed, but function b is. So now you have a lead on where to find the bug.

```python
import logging

logger = logging.getLogger(__name__)

def func_a():
  logger.debug("Executing A")
  func_b()

def func_b():
  logger.debug("Executing B")
  return  # Whoops! I "forgot" to remove this return!
  func_c()

def func_c():
logger.debug("Executing C")

def main():
  logger.debug("Executing main")
  func_a()

main()
```

hector-sab commented on metabrainz.py:
```diff
@@ -0,0 +1,72 @@
+
+
+# --------------------------------------------------
+def get_args():
+    """Get command-line arguments"""
+
+    parser = argparse.ArgumentParser(
+        description="Retrieve metadata from MusicBrainz API",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+
+    parser.add_argument(
+        "-a",
+        "--artist",
+        help="An artist to search for",
+        metavar="str",
+        type=str,
+        nargs="+",
+        required=True,
+    )
+
+    return parser.parse_args()
```
**suggestion:** This part should go in the `cli.py` file

Salv-23 replied:
Done! thanks for the suggestion.

hector-sab commented on artist_api.py:
```diff
@@ -0,0 +1,65 @@
+
+
+# --------------------------------------------------
+if __name__ == "__main__":
+    arguments = get_args()
+    artist = arguments.artist
+    main(artist)
```
**suggestion:** If this the main entry point of your program, this can/should go in `__main__.py`

Salv-23 replied:
This is just for testing purposes.

Salv-23 committed:
feat: Add instance method to gather release-groups

Salv-23 committed:
feat: Add instance method to gather releases

Salv-23 committed:
fix: Add suggested project structure

Salv-23 committed:
feat: Add methods for albums, singles and episodes

Salv-23 committed:
fix: Remove instance constructor method and switch to class methods

hector-sab commented on metamusic/clients/musicbrainz.py:
```diff
@@ -0,0 +1,126 @@
+
+    @classmethod
+    def get_artist_info(cls, artist: str) -> dict:
+        endpoint = cls.artist_endpoit
+        artist = artist.split()
+        artist = "%20".join(artist)
+        artist_query = f"{endpoint}?query={artist}&limit=3&fmt=json"
```
**suggestion:** Don't define a variable if used only once.

```suggestion
        artist = artist.split()
        artist = "%20".join(artist)
        artist_query = f"{cls.artist_endpoit}?query={artist}&limit=3&fmt=json"
```

Salv-23 replied:
You're right, fixed.

hector-sab commented on metamusic/clients/musicbrainz.py:
```diff
@@ -0,0 +1,126 @@
+
+# --------------------------------------------------
+class MusicMetadata:
+    artist_endpoit = "http://musicbrainz.org/ws/2/artist/"
+    release_groups_endpoint = "http://musicbrainz.org/ws/2/release-group/"
+    releases_endpoint = "http://musicbrainz.org/ws/2/release/"
```
**suggestion:** Don't repeat yourself.

Set a base url, you can then define the endpoint in the function that uses it or here
```suggestion
    base_url = "http://musicbrainz.org/ws/2"
```

Salv-23 replied:
You're right, thanks for this suggestion.

hector-sab commented on metamusic/clients/musicbrainz.py:
```diff
@@ -0,0 +1,126 @@
+
+    @classmethod
+    def _pick_artist(cls, artist_list):
+        for artist in artist_list:
+            if artist["score"] == 100:
+                return artist
```
**suggestion:** Set a todo or a NotImplementedError where things are pending to be done. Unless we don't want to deal with no artist and 100 score that way.

```suggestion
        for artist in artist_list:
            if artist["score"] == 100:
                return artist
        raise NotImplementedError("There's no 100 score artist. Can't handle it right now.")
```

Salv-23 committed:
fix: Remove artist info query limit

Previously the query was limited to the first three results for testing
purposes, that constraint it's no longer needed.

Salv-23 committed:
fix: Use logging library instead of pprint

Salv-23 committed:
fix: Move argparse functionality to cli.py

Salv-23 commented on metamusic/clients/musicbrainz.py:
```diff
@@ -0,0 +1,126 @@
+Purpose: Query Metadata from MusicBrainz API
+"""
+
+import argparse
```
question: Why are you so good? @hector-sab

Salv-23 committed:
fix: Include release-group data when requesting artist info

Salv-23 committed:
fix: Remove get_release_groups class method

Salv-23 committed:
fix: Refactor get_artist_info class method

Salv-23 committed:
fix: Refactor get_albums class method

Salv-23 committed:
fix: Refactor get_singles class method

Salv-23 committed:
feat: Include recordings when querying album info

Salv-23 committed:
feat: Include recordings when querying singles info

Salv-23 committed:
feat: Include recording when querying episodes info

Salv-23 committed:
fix: Change _pick_artist class method

Salv-23 committed:
fix: Create base url and entry point in __main__.py

Salv-23 commented on metamusic/clients/musicbrainz.py:
```diff
@@ -0,0 +1,121 @@
+            elif artist["score"] < posible_artist["score"]:
+                artist = posible_artist
+        if artist and artist["score"] != 100:
+            raise NotImplementedError(
+                "Selection of artists with lower score than 100 not implemented yet"
+            )
+        return artist
+
+    @classmethod
+    def _artist_lookup(cls, artist_id: str) -> dict:
+        artist_lookup = f"{cls.base_url}artist/{artist_id}?inc=release-groups&fmt=json"
```
**question**: What would be the suggested way to separate in multiple lines, from line 26 to 28 and line 33.

hector-sab replied:
For line 33, I would leave it as is, as it is a URL. Maybe use urljoin for joining cls.base_url, artist, and artist_id.

For line 27, you could do something like
```python
            raise NotImplementedError(
                "Selection of artists with lower"
                " score than 100 not implemented yet"
            )
```

Salv-23 committed:
feat: Add class methods to retrieve recording information

Salv-23 committed:
feat!: First complete release of MusicMetadata class

MusicMetaData can retrieve artist, release and title information. It
still misses complete documentation and proper handling of exceptions.

Salv-23 committed:
feat: Add dataclass for Track

Salv-23 committed:
fix: Remove shebang and imports from __main__.py

Salv-23 committed:
fix: Remove shebang and empty lines to cli.py

Salv-23 committed:
feat: Add total_tracks to Track dataclass

Salv-23 committed:
feat: Add basic setup.py

Salv-23 committed:
fix: Remove shebang and change class name to MusicBrainz

Salv-23 committed:
feat: Add eyeD3 functionality

Salv-23 committed:
fix: Remove pyproject.toml

Removing pyproject.toml fixes importing error in songmetadata.py.
TODO: add project compatible pyproject.toml

Salv-23 committed:
fix: Remove dataclass definition from songmetadata.py

Salv-23 committed:
feat: Add function to set album artwork

hector-sab commented on metamusic/clients/songmetadata.py:
```diff
@@ -0,0 +1,123 @@
+# --------------------------------------------------
+def set_selected_tags(path: str, metadata: dict, tag_selection: list[str]):
+    track = eyed3.load(path)
+    if "artist" in tag_selection:
+        track.tag.artist = metadata["artist"]
+    if "album" in tag_selection:
+        track.tag.album = metadata["album"]
+    if "title" in tag_selection:
+        track.tag.title = metadata["title"]
+    if "track_num" in tag_selection:
+        track.tag.track_num = (metadata["track_num"], metadata["total_tracks"])
+    if "album_type" in tag_selection:
+        track.tag.album_type = metadata["album_type"]
+    if "release_date" in tag_selection:
+        track.tag.original_release_date = metadata["release_date"]
```
suggestion: Use if-else or a match statement
https://docs.python.org/3/tutorial/controlflow.html#match-statements

hector-sab commented on metamusic/clients/songmetadata.py:
```diff
@@ -0,0 +1,123 @@
+
+
+# --------------------------------------------------
+def set_selected_tags(path: str, metadata: dict, tag_selection: list[str]):
```
thought: You are modifying a variable passed as argument in and outside the function without returning it.

This is purely your choice on how you want things to work, but be aware that you are modifying the metadata argument inside the function and using it outside as it was modified. This can some times get a bit confusing as regularly one expects to get a new object returned by the function. 

However you are returning a None and modifying in place the object you are interested in. This is fine, OpenCV does this for example, and I assume it does it this way for avoiding copying the object that stores image info multiple times.

Just be careful. Write some documentation (docstrings) so you remember the behavior of your function without having to dig into the code to find out.

You can read a bit more of this here
https://stackoverflow.com/questions/575196/why-can-a-function-modify-some-arguments-as-perceived-by-the-caller-but-not-oth

hector-sab replied:
Ah, nvm. I thought you were using metadata outside with some modifications done in here. Still, the problem is a little interesting and some times it´s good to know about it :)

hector-sab commented on metamusic/clients/songmetadata.py:
```diff
@@ -0,0 +1,123 @@
+import eyed3
+
+
+# --------------------------------------------------
+def get_args():
+    """Get command-line arguments"""
+
+    parser = argparse.ArgumentParser(
+        description="eyeD3 client to get or update metadata",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+
+    parser.add_argument(
+        "-t",
+        "--track",
+        help="A track file to get or update metadata",
+        metavar="str",
+        type=str,
+        required=True,
+    )
+
+    parser.add_argument(
+        "-i",
+        "--image",
+        help="An image file to update album artwork",
+        metavar="str",
+        type=str,
+        required=True,
+    )
+
+    return parser.parse_args()
```
suggestion: Move this to the cli.py file and start creating a single argparse that handles all the cases you want.

hector-sab commented on metamusic/__main__.py:
```diff
@@ -0,0 +1,5 @@
+Author : Salvador <Salvador@fedora>
+Date   : 2022-10-07
+Purpose: New program
+"""
```
suggestion: Add the entry point for your cli application.

Take a look at this small article
https://chriswarrick.com/blog/2014/09/15/python-apps-the-right-way-entry_points-and-scripts/

Salv-23 committed:
feat: Add album artwork from local file and from URL

Salv-23 committed:
chore: Add .vscode path to .gitignore

Salv-23 committed:
feat: Add url argument

Salv-23 committed:
fix: recording date on tracks

Salv-23 committed:
fix: Add proper format to songmetadata.py

Salv-23 committed:
feat: Add client to collect HD artwork

Salv-23 committed:
feat: Get release object substring

Salv-23 committed:
feat: Retrieve url for high definition artwork

Salv-23 committed:
feat: show album artwork in terminal

Salv-23 committed:
feat: remove typer functionality, add docstrings.

Salv-23 committed:
fix: restore typer functionality to artwork.py

Salv-23 committed:
fix: remove typer from artwork.py

Salv-23 committed:
fix: rename "album" to "release"

Salv-23 committed:
feat: cli first model
